### PR TITLE
GDScript LSP: Convert [color] BBCode to <span> HTML for capable clients

### LINKED ISSUE #14633: GDScript LSP: Support [color] BBCode via HTML <span> for capable clients
### Problem or limitation

The Godot built-in editor renders the `[color]` BBCode when viewing on-hover documentation. But the GDScript LSP strips it entirely because it converts to Markdown, and there is no `[color]` equivalent in Markdown. However, some LSP clients can render HTML, and a `<span style="color:...">` is functionally equivalent to `[color]`.

This proposal would positively affect any LSP client capable of rendering `<span>` HTML. The original motivation was this Rider issue: [RIDER-129495](https://youtrack.jetbrains.com/issue/RIDER-129495).

### Proposed improvement

I propose having the GDScript LSP convert the `[color]` BBCode to an equivalent `<span>` (e.g., `<span style="color: #...">`), if the client declares it supports the `<span>` HTML tag via the [MarkdownClientCapabilities](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#markdownClientCapabilities)'s `allowedTags` property.

I have prepared a [proof-of-concept](https://github.com/godotengine/godot/compare/master...Finishxx:godot:lsp-color-bbcode-support) in my branch, showcased here on a slightly modified Rider (just sending the appropriate `MarkdownClientCapabilities`). Notice that the "Warning:" text is displayed in yellow color, according to the doc comment.

Image of Rider editor displaying part of the text in on-hover documentation as being colored yellow

From my testing, it does not affect VS Code, Zed, or Vim GDScript LSP clients. They would also need to send the appropriate `MarkdownClientCapabilities` and correctly render the HTML to be affected.

### Proposal review

- [x] Yes, this proposal is highly specific and actionable.

### PR DESCRIPTION
**Pull request by Finishxx**
Closes https://github.com/godotengine/godot-proposals/issues/14633

The GDScript LSP currently strips `[color]` BBCode tags when converting documentation to Markdown, since Markdown has no equivalent. This PR preserves it for capable clients by converting `[color=...]` to `<span style="color:...">`, but only if the client declares support for it.

Specifically, conversion is applied only when `"span"` is included in `general.markdown.allowedTags`:
https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#markdownClientCapabilities

Clients that do not advertise this capability are unaffected, and the existing stripping behavior is preserved.

Tested with Rider (via https://github.com/JetBrains/godot-support/commit/476f5d36bef9f2e33c7830ff83d60179ccbaaa28), and verified that VS Code, Zed, and Vim clients are unaffected.

AI disclaimer: the original proof-of-concept in the proposal was written by AI, whilst this PR was written by hand.

### CODE REVIEW

**Comment by Finishxx** [comment]
cc: @van800 @HolonProduction

**Comment by van800** [comment]
Looks good to me, good job! 
(I haven't tested, but it seems the case with multiple simultaneous clients, which have different capabilities is supported?)

**Review by Calinou — APPROVED** [review]
Code and unit tests look good to me (didn't test).

**Comment by HolonProduction** [comment]
Will test once Rider 2026.2 is available

**Comment by van800** [comment]
2026.2 EAP is available https://www.jetbrains.com/rider/nextversion/

**Comment by Calinou** [comment]
Tested locally, it works as expected.

Godot script editor:

image

Rider 2026.2 EAP4:

image

There's a color mismatch since HTML color `green` is darker than Godot's color `green`, but it's not critical. We could replace `green` with `lime` if we wanted to get matching colors.

As an aside, I noticed Rider on Linux thinks the project path is mismatched if it's a symlinked folder (both the Godot editor and Rider project use the same symlinked folder path). I had to open Rider and Godot on the actual folder for the paths to actually be considered matching, and the LSP to be functional.

**Review by bruvzg — APPROVED** [review]

File: modules/gdscript/language_server/gdscript_language_protocol.cpp [hunk 1]
```diff
@@ -276,6 +276,13 @@ Variant GDScriptLanguageProtocol::initialize(const Dictionary &p_params) {
 	client->behavior.use_snippets_for_brace_completion = get_deep(capabilities, false,
 			"textDocument", "completion", "completionItem", "snippetSupport");
 
+	Array allowed_tags = get_deep(capabilities, Array(), "general", "markdown", "allowedTags");
+	for (int i = 0; i < allowed_tags.size(); i++) {
+		if (allowed_tags[i].get_type() == Variant::STRING) {
+			client->behavior.markdown_allowed_html_tags.insert(allowed_tags[i]);
+		}
+	}
```
Comment (HolonProduction) [inline modules/gdscript/language_server/gdscript_language_protocol.cpp:284]:
```suggestion
	Array allowed_tags = get_deep(capabilities, Array(), "general", "markdown", "allowedTags");
	for (const Variant &tag : allowed_tags) {
		if (tag.is_string()) {
			client->behavior.markdown_allowed_html_tags.insert(tag);
		}
	}
```

File: modules/gdscript/language_server/gdscript_language_protocol.h [hunk 2]
```diff
@@ -167,6 +169,12 @@ class GDScriptLanguageProtocol : public JSONRPC {
 	 */
 	ExtendGDScriptParser *get_parse_result(const String &p_path);
 
+	/**
+	 * Returns the HTML tags the client can render inside Markdown content, e.g. `span`.
+	 * Defaults to an empty set if no client is connected.
+	 */
+	const HashSet<String> &get_client_markdown_allowed_html_tags() const;
```
Comment (HolonProduction) [inline modules/gdscript/language_server/gdscript_language_protocol.h:178]:
```suggestion
	/**
	 * Returns the HTML tags the client can render inside Markdown content, e.g. `span`.
	 * Defaults to an empty set if no client is connected.
	 * 
	 * TODO: Remove after moving endpoints into unified class. Access non-null client directly.
	 */
	const HashSet<String> &get_client_markdown_allowed_html_tags() const;
```

File: modules/gdscript/language_server/gdscript_language_protocol.h [hunk 3]
```diff
@@ -53,6 +53,8 @@ class GDScriptLanguageProtocol : public JSONRPC {
 	struct ClientBehavior {
 		/** If `true` use snippet insert mode to position the cursor between braces of completion options. If `false` strip braces from completion options since we can't provide good UX for them. */
 		bool use_snippets_for_brace_completion = false;
+		/** HTML tags, e.g. `span`, which the client declares is capable of rendering inside Markdown content. */
```
Comment (HolonProduction) [inline modules/gdscript/language_server/gdscript_language_protocol.h]:
```suggestion
		/** HTML tags, e.g. `span`, which the client declares is capable of rendering inside Markdown content. Assumes full support for attributes e.g. `style`. */
```

File: modules/gdscript/language_server/godot_lsp.h [hunk 4]
```diff
@@ -1299,13 +1299,13 @@ struct DocumentSymbol {
 		return dict;
 	}
 
-	_FORCE_INLINE_ MarkupContent render() const {
+	_FORCE_INLINE_ MarkupContent render(const HashSet<String> &p_allowed_html_tags = {}) const {
```
Comment (HolonProduction) [inline modules/gdscript/language_server/godot_lsp.h]:
```suggestion
	_FORCE_INLINE_ MarkupContent render(const HashSet<String> &p_allowed_html_tags) const {
```

File: modules/gdscript/language_server/godot_lsp.h [hunk 5]
```diff
@@ -1909,7 +1909,8 @@ struct GodotCapabilities {
 };
 
 /** Format BBCode documentation from DocData to markdown */
-static String marked_documentation(const String &p_bbcode) {
+static String marked_documentation(const String &p_bbcode, const HashSet<String> &p_allowed_html_tags = {}) {
```
Comment (HolonProduction) [inline modules/gdscript/language_server/godot_lsp.h]:
```suggestion
static String marked_documentation(const String &p_bbcode, const HashSet<String> &p_allowed_html_tags) {
```

File: modules/gdscript/tests/test_lsp.h [hunk 6]
```diff
@@ -553,11 +553,22 @@ func f():
...
+		CHECK_EQ(LSP::marked_documentation("[color=red]red text[/color]"), "red text");
+
+		// [color] tag for capable clients is converted to span
+		HashSet<String> span_allowed_tags = { "span" };
+		CHECK_EQ(LSP::marked_documentation("[color=red]red text[/color]", span_allowed_tags), "<span style=\"color:red\">red text</span>");
+		CHECK_EQ(LSP::marked_documentation("[color=#ff0000]red text[/color]", span_allowed_tags), "<span style=\"color:#ff0000\">red text</span>");
+		CHECK_EQ(LSP::marked_documentation("[color=red]r[/color] and [color=blue]b[/color]", span_allowed_tags), "<span style=\"color:red\">r</span> and <span style=\"color:blue\">b</span>");
+		CHECK_EQ(LSP::marked_documentation("[color=lobster]invalid color name[/color]", span_allowed_tags), "invalid color name");
+		CHECK_EQ(LSP::marked_documentation("[color=#zzzzzz]invalid color hex[/color]", span_allowed_tags), "invalid color hex");
+		CHECK_EQ(LSP::marked_documentation("[color=#ff0000\"> <script>malicious code</script>]no escaping[/color]", span_allowed_tags), "no escaping");
```
Comment (HolonProduction) [inline modules/gdscript/tests/test_lsp.h:571]: A case with alpha component would make sense.

**Review by HolonProduction — COMMENTED** [review]
Alpha seems to not be displayed in Rider. But looking at the generated markdown, this seems to be a thing with Riders MD renderer not an issue with this code.

Also in theory color tags can be nested:

Image

which this PR does not properly transform.

Image

I'm a bit conflicted on this. I think we ignored a similar nesting edge case with URLs, so it should be fine I guess.

File: modules/gdscript/tests/test_lsp.h (hunk 6 above)
Comment (Finishxx) [reply modules/gdscript/tests/test_lsp.h:571]: Good call, added the test case

**Comment by HolonProduction** [comment]
> There's a color mismatch since HTML color `green` is darker than Godot's color `green`, but it's not critical. We could replace `green` with `lime` if we wanted to get matching colors.

Coming back to this for a second. Why keep color names in the first place? Wouldn't it make more sense to convert all colors to html notation using [`Color::from_string`](https://docs.godotengine.org/en/stable/classes/class_color.html#class-color-method-from-string)?

**Comment by Finishxx** [comment]
> > There's a color mismatch since HTML color `green` is darker than Godot's color `green`, but it's not critical. We could replace `green` with `lime` if we wanted to get matching colors.
> 
> Coming back to this for a second. Why keep color names in the first place? Wouldn't it make more sense to convert all colors to html notation using [`Color::from_string`](https://docs.godotengine.org/en/stable/classes/class_color.html#class-color-method-from-string)?

I like that idea - keeps the colors consistent with Godot's built-in editor across the IDEs. Fixed in the last push

Rider' green is now same as Godot's:
Screenshot_2026-06-15_16-22-26

File: modules/gdscript/language_server/gdscript_language_protocol.h [hunk 7]
```diff
@@ -53,6 +53,8 @@ class GDScriptLanguageProtocol : public JSONRPC {
 	struct ClientBehavior {
 		/** If `true` use snippet insert mode to position the cursor between braces of completion options. If `false` strip braces from completion options since we can't provide good UX for them. */
 		bool use_snippets_for_brace_completion = false;
+		/** HTML tags, e.g. `span`, which the client declares is capable of rendering inside Markdown content. Assumes full support for attributes e.g. `style`. */
```
Comment (HolonProduction) [inline modules/gdscript/language_server/gdscript_language_protocol.h]:
```suggestion
		/** HTML tags, e.g. `span`, which the client is capable of rendering inside Markdown content. Assumes full support for attributes e.g. `style`. */
```

**Review by HolonProduction — APPROVED** [review]
LGTM

I'm a bit concerned about the nested color thing. But I think we can fix that on its own, if it actually comes up for users.

**Comment by Finishxx** [comment]
> I'm a bit concerned about the nested color thing. But I think we can fix that on its own, if it actually comes up for users.

I think the concern is justified, as this is a clear deviation from how the built-in editor displays it. I checked, and the [url] tag handling in the LSP currently has the same limitation as [color]. 

Couldn't find anyone bothered by the [url] not handling nesting, so I also think it should be fine to leave [color] like this, too. If [color] were to handle nesting, it should definitely use the same apparatus (most likely a [stack of tags](https://github.com/godotengine/godot/blob/c309339ec6b3d1c17ef9e49b27306e30e721a182/editor/doc/editor_help.cpp#L2534) would be involved) for [url] as well, which feels out of scope for this PR.

**Comment by Repiteo** [comment]
Thanks! Congratulations on your first merged contribution! 🎉
